transport/server.cc

```diff
@@ -976,12 +976,12 @@ template <typename Process>
                                    cql3::computed_function_values,
                                    cql3::dialect>
 future<cql_server::result_with_foreign_response_ptr>
-cql_server::connection::process_on_shard(::shared_ptr<messages::result_message::bounce_to_shard> bounce_msg, uint16_t stream, fragmented_temporary_buffer::istream is,
-        service::client_state& cs, service_permit permit, tracing::trace_state_ptr trace_state, cql3::dialect dialect, Process process_fn) {
-    return _server.container().invoke_on(*bounce_msg->move_to_shard(), _server._config.bounce_request_smp_service_group,
+cql_server::connection::process_on_shard(shard_id shard, uint16_t stream, fragmented_temporary_buffer::istream is,
+        service::client_state& cs, service_permit permit, tracing::trace_state_ptr trace_state, cql3::dialect dialect, cql3::computed_function_values&& cached_vals, Process process_fn) {
+    return _server.container().invoke_on(shard, _server._config.bounce_request_smp_service_group,
             [this, is = std::move(is), cs = cs.move_to_other_shard(), stream, permit = std::move(permit), process_fn,
              gt = tracing::global_trace_state_ptr(std::move(trace_state)),
-             cached_vals = std::move(bounce_msg->take_cached_pk_function_calls()), dialect] (cql_server& server) {
+             cached_vals = std::move(cached_vals), dialect] (cql_server& server) mutable {
```

Comment:
Good catch. I guess the limitation is because of the single-threaded custom allocator?

Reply:
> Good catch. I guess the limitation is because of the single-threaded custom allocator?

Yes. We do have cross shard free (mostly because of exception which we do not control where they are allocated), but it is less efficient (it is a single consumer multiple producers queue with all the locking that it entails).